File: Data Structures & Algorithms/sliding-window-maximum/submission-4.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        // will store pair of ints, where [0] is the value and [1] is the index
        Deque<int[]> deque = new LinkedList<>();

        List<Integer> list = new ArrayList<>();

        // create first window
        // for (int i = 0; i < k; i++) {
        //     maxHeap.add(new int[] { nums[i], i });
        // }

        int left = 0;
        // start right r at the actual right index (inclusive) and add to maxHeap first
        for (int r = 0; r < nums.length; r++) {
            while (!deque.isEmpty() && deque.peekLast()[0] < nums[r]) {
                deque.pollLast();
            }
            deque.addLast(new int[] { nums[r], r });


            if (left > deque.peekFirst()[1]) {
                deque.pollFirst();
            }

            // move sliding window over one if applicable
            // add to result
            if ((r + 1) >= k) {
                list.add(deque.peekFirst()[0]);
                left++;
            }
            
        }

        return list.stream().mapToInt(i -> i).toArray();
    }
}

File: Data Structures & Algorithms/palindromic-substrings/submission-4.js
class Solution {
    /**
     * @param {string} s
     * @return {number}
     */
    countSubstrings(s) {
        if (!s) {
            return 0;
        }

        let result = 0;
        for (let i = 0; i < s.length; i++) {
            result += this.countPalindromes(s, i, true);
            result += this.countPalindromes(s, i, false);
        }

        return result;
    }

    countPalindromes(s, index, isOdd) {
        // get all palindromes where center is at index

        let left = index;
        let right = isOdd ? index : index + 1;

        let result = 0;
        while (left >= 0 && right < s.length && s.at(left) === s.at(right)) {
            result++;
            left--;
            right++;
        }

        return result;
    }
}
